File: android/src/main/java/me/carda/awesome_notifications/awesome_notifications_core/broadcasters/senders/BroadcastSender.java
package me.carda.awesome_notifications.awesome_notifications_core.broadcasters.senders;

import android.content.Context;
import android.content.Intent;

import androidx.core.app.JobIntentService;
import me.carda.awesome_notifications.awesome_notifications_core.Definitions;
import me.carda.awesome_notifications.awesome_notifications_core.builders.NotificationBuilder;
import me.carda.awesome_notifications.awesome_notifications_core.broadcasters.receivers.AwesomeEventsReceiver;
import me.carda.awesome_notifications.awesome_notifications_core.enumerators.NotificationLifeCycle;
import me.carda.awesome_notifications.awesome_notifications_core.managers.ActionManager;
import me.carda.awesome_notifications.awesome_notifications_core.managers.DismissedManager;
import me.carda.awesome_notifications.awesome_notifications_core.managers.CreatedManager;
import me.carda.awesome_notifications.awesome_notifications_core.managers.DisplayedManager;
import me.carda.awesome_notifications.awesome_notifications_core.managers.LifeCycleManager;
import me.carda.awesome_notifications.awesome_notifications_core.models.returnedData.ActionReceived;
import me.carda.awesome_notifications.awesome_notifications_core.models.returnedData.NotificationReceived;
import me.carda.awesome_notifications.awesome_notifications_core.services.BackgroundService;

public class BroadcastSender {

    private static final String TAG = "BroadcastSender";

    public static void sendBroadcastNotificationCreated(Context context, NotificationReceived notificationReceived){

        if (notificationReceived != null)
            try {

                if(LifeCycleManager.getApplicationLifeCycle() == NotificationLifeCycle.AppKilled) {
                    CreatedManager.saveCreated(context, notificationReceived);
                    CreatedManager.commitChanges(context);
                }
                else {
                    AwesomeEventsReceiver
                            .getInstance()
                            .addNotificationEvent(
                                    context,
                                    Definitions.BROADCAST_CREATED_NOTIFICATION,
                                    notificationReceived);
                }

            } catch (Exception e) {
                e.printStackTrace();
            }
    }

    public static void sendBroadcastNotificationDisplayed(Context context, NotificationReceived notificationReceived){

        if (notificationReceived != null)
            try {

                if(LifeCycleManager.getApplicationLifeCycle() == NotificationLifeCycle.AppKilled) {
                    DisplayedManager.saveDisplayed(context, notificationReceived);
                    DisplayedManager.commitChanges(context);
                }
                else {
                    AwesomeEventsReceiver
                            .getInstance()
                            .addNotificationEvent(
                                    context,
                                    Definitions.BROADCAST_DISPLAYED_NOTIFICATION,
                                    notificationReceived);
                }

            } catch (Exception e) {
                e.printStackTrace();
            }
    }

    public static void sendBroadcastNotificationDismissed(Context context, ActionReceived actionReceived){
        if (actionReceived != null)
            try {

                if(LifeCycleManager.getApplicationLifeCycle() == NotificationLifeCycle.AppKilled) {
                    DismissedManager.saveDismissed(context, actionReceived);
                    DismissedManager.commitChanges(context);
                }
                else {
                    AwesomeEventsReceiver
                            .getInstance()
                            .addAwesomeActionEvent(
                                    context,
                                    Definitions.BROADCAST_DISMISSED_NOTIFICATION,
                                    actionReceived);
                }

            } catch (Exception e) {
                e.printStackTrace();
            }
    }

    public static void sendBroadcastDefaultAction(Context context, ActionReceived actionReceived){
        if (actionReceived != null)
            try {

                if(LifeCycleManager.getApplicationLifeCycle() == NotificationLifeCycle.AppKilled) {
                    ActionManager.saveAction(context, actionReceived);
                    ActionManager.commitChanges(context);
                }
                else {
                    AwesomeEventsReceiver
                            .getInstance()
                            .addAwesomeActionEvent(
                                    context,
                                    Definitions.BROADCAST_DEFAULT_ACTION,
                                    actionReceived);
                }

            } catch (Exception e) {
                e.printStackTrace();
            }
    }

    public static void sendBroadcastSilentAction(Context context, ActionReceived actionReceived){
        if (actionReceived != null)
            try {
                AwesomeEventsReceiver
                        .getInstance()
                        .addAwesomeActionEvent(
                                context,
                                Definitions.BROADCAST_SILENT_ACTION,
                                actionReceived);

            } catch (Exception e) {
                e.printStackTrace();
            }
    }

    public static void sendBroadcastBackgroundAction(Context context, ActionReceived actionReceived){

        if (actionReceived != null)
            try {
                AwesomeEventsReceiver
                        .getInstance()
                        .addAwesomeActionEvent(
                                context,
                                Definitions.BROADCAST_BACKGROUND_ACTION,
                                actionReceived);

            } catch (Exception e) {
                e.printStackTrace();
            }
    }

    public static void enqueueSilentAction(Context context, String previousAction, ActionReceived actionReceived){

        Intent serviceIntent =
                NotificationBuilder
                        .getNewBuilder()
                        .buildNotificationIntentFromActionModel(
                                context,
                                previousAction,
                                actionReceived,
                                BackgroundService.class);

        JobIntentService.enqueueWork(
                context,
                BackgroundService.class,
                42,
                serviceIntent);
    }

    public static void enqueueSilentBackgroundAction(Context context, String previousAction, ActionReceived actionReceived){

        Intent serviceIntent =
                NotificationBuilder
                        .getNewBuilder()
                        .buildNotificationIntentFromActionModel(
                                context,
                                previousAction,
                                actionReceived,
                                BackgroundService.class);

        JobIntentService.enqueueWork(
                context,
                BackgroundService.class,
                42,
                serviceIntent);
    }
}

File: ios/core_old/broadcasters/senders/BroadcastSender.swift
//
//  BroadcastSender.swift
//  awesome_notifications
//
//  Created by CardaDev on 02/02/22.
//

import Foundation

class BroadcastSender {
    
    private let TAG = "BroadcastSender"
    
    // ************** SINGLETON PATTERN ***********************
    
    static var instance:BroadcastSender?
    public static var shared:BroadcastSender {
        get {
            BroadcastSender.instance =
                BroadcastSender.instance ?? BroadcastSender()
            return BroadcastSender.instance!
        }
    }
    private init(){}
    
    // ********************************************************
    
    public func sendBroadcast(
        notificationCreated notificationReceived: NotificationReceived,
        whenFinished completionHandler: @escaping (Bool) -> Void
    ){
        if SwiftUtils.isRunningOnExtension() || LifeCycleManager.shared.currentLifeCycle == .AppKilled {
            CreatedManager.saveCreated(received: notificationReceived)
        }
        else {
            AwesomeEventsReceiver
                .shared
                .addNotificationEvent(
                    named: Definitions.BROADCAST_CREATED_NOTIFICATION,
                    with: notificationReceived)
        }
        
        completionHandler(true)
    }
    
    public func sendBroadcast(
        notificationDisplayed notificationReceived: NotificationReceived,
        whenFinished completionHandler: @escaping (Bool) -> Void
    ){
        if SwiftUtils.isRunningOnExtension() || LifeCycleManager.shared.currentLifeCycle == .AppKilled {
            DisplayedManager.saveDisplayed(received: notificationReceived)
        }
        else {
            AwesomeEventsReceiver
                .shared
                .addNotificationEvent(
                    named: Definitions.BROADCAST_DISPLAYED_NOTIFICATION,
                    with: notificationReceived)
        }
        
        DefaultsManager
            .shared
            .registerLastDisplayedDate()
        
        completionHandler(true)
    }
    
    public func sendBroadcast(
        actionReceived: ActionReceived,
        whenFinished completionHandler: @escaping (Bool, Error?) -> Void
    ){
        if !ActionManager.recovered { //LifeCycleManager.shared.currentLifeCycle == .AppKilled
            ActionManager.saveAction(received: actionReceived)
            Logger.i(TAG, "action saved")
        }
        else {
            AwesomeEventsReceiver
                .shared
                .addActionEvent(
                    named: Definitions.BROADCAST_DEFAULT_ACTION,
                    with: actionReceived)
            Logger.i(TAG, "action broadcasted")
        }
        
        completionHandler(true, nil)
    }
    
    public func sendBroadcast(
        notificationDismissed actionReceived: ActionReceived,
        whenFinished completionHandler: @escaping (Bool, Error?) -> Void
    ){
        if LifeCycleManager.shared.currentLifeCycle == .AppKilled {
            DismissedManager.saveDismissed(received: actionReceived)
        }
        else {
            AwesomeEventsReceiver
                .shared
                .addActionEvent(
                    named: Definitions.BROADCAST_DISMISSED_NOTIFICATION,
                    with: actionReceived)
        }
        
        completionHandler(true, nil)
    }
    
    public func sendBroadcast(
        silentAction: ActionReceived,
        whenFinished completionHandler: @escaping (Bool, Error?) -> Void
    ){
        AwesomeEventsReceiver
            .shared
            .addActionEvent(
                named: Definitions.BROADCAST_SILENT_ACTION,
                with: silentAction)
        
        completionHandler(true, nil)
    }
    
    public func sendBroadcast(
        backgroundAction: ActionReceived,
        whenFinished completionHandler: @escaping (Bool, Error?) -> Void
    ){
        AwesomeEventsReceiver
            .shared
            .addActionEvent(
                named: Definitions.BROADCAST_BACKGROUND_ACTION,
                with: backgroundAction)
        
        completionHandler(true, nil)
    }
    
    public func enqueue(
        silentAction actionReceived: ActionReceived,
        whenFinished completionHandler: @escaping (Bool, Error?) -> Void
    ){
        BackgroundService
            .shared
            .enqueue(
                SilentBackgroundAction: actionReceived,
                withCompletionHandler: completionHandler)
    }
    
    public func enqueue(
        silentBackgroundAction actionReceived: ActionReceived,
        whenFinished completionHandler: @escaping (Bool, Error?) -> Void
    ){
        BackgroundService
            .shared
            .enqueue(
                SilentBackgroundAction: actionReceived,
                withCompletionHandler: completionHandler)
    }
    
}
